inventory_management_system_api/repositories/setting.py

```diff
@@ -0,0 +1,107 @@
+"""
+Module for providing a repository for managing settings in a MongoDB database.
+"""
+
+import logging
+from typing import Optional, Type, TypeVar
+
+from pymongo.client_session import ClientSession
+from pymongo.collection import Collection
+
+from inventory_management_system_api.core.database import DatabaseDep
+from inventory_management_system_api.models.setting import SettingInBase, SettingOutBase, SparesDefinitionOut
+
+logger = logging.getLogger()
+
+# Template types for models inheriting from SettingIn/OutBase so this repo can be used generically for multiple settings
+SettingInBaseT = TypeVar("SettingInBaseT", bound=SettingInBase)
+SettingOutBaseT = TypeVar("SettingOutBaseT", bound=SettingOutBase)
+
+
+# Aggregation pipeline for getting the spares definition complete with usage status data
+SPARES_DEFINITION_GET_AGGREGATION_PIPELINE: list = [
+    # Only perform this on the relevant document
+    {"$match": {"_id": SparesDefinitionOut.SETTING_ID}},
+    # Deconstruct the usage statuses so can go through them one by one
+    {"$unwind": "$usage_statuses"},
+    # Find and store actual usage status data as 'statusDetails'
+    {
+        "$lookup": {
+            "from": "usage_statuses",
+            "localField": "usage_statuses.id",
+            "foreignField": "_id",
+            "as": "statusDetails",
+        }
+    },
+    {"$unwind": "$statusDetails"},
+    # Merge the two sets of documents together
+    {"$addFields": {"usage_statuses": {"$mergeObjects": ["$usage_statuses", "$statusDetails"]}}},
+    # Remove the temporary 'statusDetails' field as no longer needed
+    {"$unset": "statusDetails"},
+    # Reconstruct the original document by merging with the original fields
+    {
+        "$group": {
+            "_id": "$_id",
+            "usage_statuses": {"$push": "$usage_statuses"},
+            "otherFields": {"$first": "$$ROOT"},
+        }
+    },
+    {"$replaceRoot": {"newRoot": {"$mergeObjects": ["$otherFields", {"usage_statuses": "$usage_statuses"}]}}},
+]
+
+
+class SettingRepo:
+    """
+    Repository for managing settings in a MongoDB database.
+    """
+
+    def __init__(self, database: DatabaseDep) -> None:
+        """
+        Initialize the `SettingRepo` with a MongoDB database instance.
+
+        :param database: The database to use.
+        """
+        self._database = database
+        self._settings_collection: Collection = self._database.settings
+
+    def upsert(
+        self, setting: SettingInBaseT, out_model_type: Type[SettingOutBaseT], session: ClientSession = None
+    ) -> SettingOutBaseT:
```

Comment:
```suggestion
    def upsert(
        self, setting: SettingInBaseT, out_model_type: Type[SettingOutBaseT], session: ClientSession = None
    ) -> Optional[SettingOutBaseT]:
```
The `get` used in the return of this function has an optional return type. We do expect a `setting` since we just upserted one previously, but technically by the get function's defintion, the return value could be `None`, so I'm not sure whether an Optional is necessary or not for the return type.

Reply:
Yeah I opted for not having it because the get in this case comes after the upsert so it should always exist in this context. I believe we do the same on the other updates for the same reason.